iceoryx_dust/vocabulary/include/iox/span.hpp

```diff
@@ -0,0 +1,429 @@
+    constexpr explicit span(const span<U, OtherExtent>& other);
+
+    constexpr span& operator=(const span&) noexcept = default;
+
+    constexpr span(span&& other) noexcept;
+    constexpr span& operator=(span&& other) noexcept;
+
+    ~span() noexcept = default;
+
+    // subviews
+
+    /// @brief obtains a subspan consisting of the first N elements of the sequence
+    /// @tparam Count number of elements in the returned span
+    /// @return the subspan
+    template <uint64_t Count>
+    constexpr span<T, Count> first() const noexcept;
+
+    /// @brief obtains a subspan consisting of the first N elements of the sequence
+    /// @param count number of elements in the returned span
+    /// @return the subspan
+    constexpr span<T, DYNAMIC_EXTENT> first(uint64_t count) const noexcept;
+
+    /// @brief obtains a subspan consisting of the last N elements of the sequence
+    /// @tparam Count number of elements in the returned span
+    /// @return the subspan
+    template <uint64_t Count>
+    constexpr span<T, Count> last() const noexcept;
+
+
+    /// @brief obtains a subspan consisting of the last N elements of the sequence
+    /// @param count number of elements in the returned span
+    /// @return the subspan
+    constexpr span<T, DYNAMIC_EXTENT> last(uint64_t count) const noexcept;
+
+    /// @brief obtains a subspan
+    /// @tparam Offset in the returned span
+    /// @tparam Count number of elements in the returned span
+    /// @return the subspan
+    template <uint64_t Offset, uint64_t Count = DYNAMIC_EXTENT>
+    constexpr span<T, (Count != DYNAMIC_EXTENT ? Count : (Extent != DYNAMIC_EXTENT ? Extent - Offset : DYNAMIC_EXTENT))>
+    subspan() const noexcept;
+
+    /// @brief obtains a subspan with dynamic extend
+    /// @param offset in the returned span
+    /// @param count number of elements in the returned span
+    /// @return the subspan
+    constexpr span<T, DYNAMIC_EXTENT> subspan(uint64_t offset, uint64_t count = DYNAMIC_EXTENT) const noexcept;
+
+    // observers
+
+    /// @brief returns the number of elements in the sequence
+    /// @return number of elements
+    constexpr uint64_t size() const noexcept;
+
+    /// @brief returns the size of the sequence in bytes
+    /// @return size in bytes
+    constexpr uint64_t size_bytes() const noexcept;
+
+    /// @brief checks if the sequence is empty
+    /// @return true if empty, otherwise false
+    constexpr bool empty() const noexcept;
+
+    // element access
+
+    /// @brief access the first element
+    /// @return reference to first element
+    constexpr T& front() const noexcept;
+
+    /// @brief access the last element
+    /// @return reference to first element
+    constexpr T& back() const noexcept;
+
+    /// @brief accesses an element of the sequence
+    /// @param index which should be accessed
+    /// @return reference to respective element
+    constexpr T& operator[](uint64_t index) const noexcept;
```

Comment:
Not part of the STL but since other container have an `at` which is also bounds checked, should we add such a method. Follow up question, should we have an `unchecked_at` where the bounds check is omitted but the user has to ensure that the access is within bounds? This would be useful for code in the hot path.

Reply:
I think you are mistaken `operator[]` is [part of the STL](https://en.cppreference.com/w/cpp/container/span/operator_at).

Regarding

1. `at`: `std::span` does not have this method, so I would not add it either
2. `at_unchecked`: I would not add it right away due to the principle `Always put safety over performance`

Reply:
Okay, my comment could be mistaken. I meant that `at` is not part of the STL :sweat_smile: 

Well, due to `Always put safety over performance` we have bound checks by default for `operator[]`. There are already a few container who have an `unchecked_at` ... or maybe `at_unchecked`. No strong feelings here though. Can also be added later on.

Reply:
Let's add it later.